clients/python/samsara_sdk/client.py

```diff
@@ -0,0 +1,326 @@
+        # default 30s
+        "send_timeout": 30,
+
+        # whether of not the payload should be compressed
+        # allowed values {"gzip", "none"}
+        "compression": COMPRESSION.GZIP
+
+        TODO: support this:
+        # add samsara client statistics events
+        # this helps you to understand whether the
+        # buffer size and publish-intervals are
+        # adequately configured.
+        # "send_client_stats": True
+        """
+        self._init_config(dict(DEFAULT_CONFIG, **config))
+
+        if self.config['start_publishing_thread']:
+            self.start_consuming()
+
+    def _init_config(self, config):
+        """
+        Validate and sanitize client configuration, initalize buffer/publisher
+        """
+        if not config['url']:
+            raise InvalidConfiguration(
+                "Missing Samsara's ingestion api endpoint url.")
+        if config['publish_interval'] < 0:
+            raise InvalidConfiguration(
+                "Publish interval needs to be above 0, not {}"
+                .format(config['publish_interval']))
+        if config['min_buffer_size'] > config['max_buffer_size'] \
+           or config['min_buffer_size'] == 0:
+            config['min_buffer_size'] = 1
+
+        self.buffer = deque(maxlen=config['max_buffer_size'])
+        self.publisher = None
+        self.config = config
+
+    def start_consuming(self):
+        """
+        Stops existing publisher thread and
+        initializes a thread publish buffered at config["publish_interval"]
+        """
+        if self.publisher:
+            self.stop_consuming()
+        self.publisher = self._get_timer(self.config['publish_interval'])
+        logging.debug(
+            'Starting publishing thread at an {}sec interval'
+            .format(self.config['publish_interval']))
+        self.publisher.start()
+
+    def stop_consuming(self):
+        """Stops existing the publisher thread"""
+        if self.publisher:
+            self.publisher.stop()
+
+    def _flush_buffer(self):
+        """
+        Grabs events from the buffer and attempts to publish them
+        if the events were not successfully published
+        it requeues them
+        """
+        # popleft is atomic and allows the local thread with safe
+        # access to the elements
+        events = [self.buffer.popleft() for _ in range(len(self.buffer))]
+        success = self.publish_events(events)
+        count = len(events)
+        if success:
+            logging.debug(
+                'Successfully submitted {} events'
+                .format(count))
+        else:
+            logging.error(
+                '{} events were not successfully submitted, requeueing'
+                .format(count))
+            # if we were unable to publish the events
+            # we add them to buffer and hope to publish them
+            # later
+            for e in events:
+                self.record_event(e)
```

Comment:
Hi,
this can cause quite severe issues.
One thing we need to guarantee is the ordering of events.
by dequeuing the events and re-enqueuing them you lose the relative order of events.

We need to find a better strategy. Check if you can peek to the events in the queue without removing them from the buffer. You can then remove them once you get a successful response from the API.